File: library/src/main/java/com/a65apps/library/presenters/PersonListPresenter.java
package com.a65apps.library.presenters;

import androidx.annotation.NonNull;


import com.a65apps.core.interactors.persons.PersonListInteractor;
import com.a65apps.library.mapper.PersonModelCompactDataMapper;
import com.a65apps.library.views.PersonListView;

import com.arellomobile.mvp.InjectViewState;
import com.arellomobile.mvp.MvpPresenter;

import java.util.concurrent.TimeUnit;

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.disposables.CompositeDisposable;
import io.reactivex.rxjava3.schedulers.Schedulers;
import io.reactivex.rxjava3.subjects.PublishSubject;

@InjectViewState
public class PersonListPresenter extends MvpPresenter<PersonListView> {
    @NonNull
    private final PersonListInteractor personListInteractor;

    @NonNull
    private PersonModelCompactDataMapper dataMapper;

    @NonNull
    private CompositeDisposable compositeDisposable;

    @NonNull
    private PublishSubject<String> publishSubject;

    public PersonListPresenter(@NonNull PersonListInteractor personListInteractor) {
        this.personListInteractor = personListInteractor;
        this.dataMapper = new PersonModelCompactDataMapper();
        this.compositeDisposable = new CompositeDisposable();
        this.publishSubject = PublishSubject.create();
        compositeDisposable.add(
                publishSubject.debounce(400, TimeUnit.MILLISECONDS, Schedulers.io())
                        .switchMapSingle(searchString -> personListInteractor.loadAllPersons(searchString))
                        .observeOn(AndroidSchedulers.mainThread())
                        .doOnSubscribe(x -> getViewState().showProgressBar())
                        .subscribe(
                                contactList -> {

                                    getViewState().fetchContactList(dataMapper.transform(contactList));
                                    getViewState().hideProgressBar();
                                },
                                e -> {
                                    getViewState().fetchError(e.getMessage());
                                    getViewState().hideProgressBar();
                                }
                        )
        );
    }

    public void requestContactList(@NonNull String searchString) {
        publishSubject.onNext(searchString);
    }

    @Override
    public void onDestroy() {
        this.compositeDisposable.dispose();
        this.dataMapper = null;
        super.onDestroy();
    }
}

File: library/src/main/java/com/a65apps/library/presenters/PersonListPresenter.kt
package com.a65apps.library.presenters

import android.util.Log
import com.a65apps.core.interactors.persons.PersonListInteractor
import com.a65apps.library.mapper.PersonModelCompactDataMapper
import com.a65apps.library.views.PersonListView
import com.arellomobile.mvp.InjectViewState
import com.arellomobile.mvp.MvpPresenter
import kotlinx.coroutines.CoroutineExceptionHandler
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.channels.BroadcastChannel
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.asFlow
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.debounce
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

private const val LOG_TAG = "person_presenter"

@InjectViewState
class PersonListPresenter(
        private val personListInteractor: PersonListInteractor,
        private val dataMapper: PersonModelCompactDataMapper) : MvpPresenter<PersonListView>() {
    private val handler = CoroutineExceptionHandler { _, exception ->
        exception.message?.let {
            viewState.fetchError(it)
        }
        viewState.hideProgressBar()
    }
    private val job = SupervisorJob()
    private var scope: CoroutineScope = CoroutineScope(Dispatchers.Main + job + handler)
    private val broadcastChannel = BroadcastChannel<String>(Channel.CONFLATED)

    init {
        scope.launch {
            broadcastChannel.asFlow()
                    .debounce(400)
                    .onEach {
                        withContext(Dispatchers.Main) { viewState.showProgressBar() }
                    }
                    .flatMapLatest {
                        Log.d(LOG_TAG, "Запрашиваем данные из репозитория $it")
                        personListInteractor.loadAllPersonsFlow(it)
                    }.flowOn(Dispatchers.IO)
                    .distinctUntilChanged()
                    .collect { list ->
                        Log.d(LOG_TAG, "Обновление списка ${list.size}")
                        viewState.fetchContactList(dataMapper.transform(list))
                        viewState.hideProgressBar()
                        Log.d(LOG_TAG, "Скрыли прогрессбар")
                    }
        }
    }

    private fun loadPersonList(searchString: String) {
        scope.launch {
            viewState.showProgressBar()
            val listPersons = personListInteractor.loadAllPersons(searchString)
            viewState.fetchContactList(dataMapper.transform(listPersons))
        }
    }

    fun requestContactList(searchString: String) {
        Log.d(LOG_TAG, "Выполняем поиск по $searchString")
        broadcastChannel.offer(searchString)
    }

    override fun onDestroy() {
        Log.d(LOG_TAG, "Flow завершен")
        broadcastChannel.close()
        scope.cancel()
        super.onDestroy()
    }
}
